monai/application/datasets.py

```diff
@@ -0,0 +1,114 @@
+# Copyright 2020 MONAI Consortium
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import sys
+import tarfile
+import numpy as np
+from typing import Any, Callable
+from monai.data import CacheDataset
+from monai.transforms import Randomizable
+from .utils import download_url, extractall
+
+
+class MedNISTDataset(Randomizable, CacheDataset):
+    """
+    The Dataset to automatically download MedNIST data and generate items for training, validation or test.
+    It's based on `CacheDataset` to accelerate the training process.
+
+    Args:
+        root: target dictionary to download and load MedNIST dataset.
+        section: expected data section, can be: `training`, `validation` or `test`.
+        transform: transforms to execute operations on input data.
+        download: whether to download the MedNIST from resource link, default is False.
+            if expected file already exists, skip downloading even set it to True.
+            user can manually copy `MedNIST.tar.gz` file or `MedNIST` folder to root directory.
+        extract: whether to extract the MedNIST.tar.gz file under root directory, default is False.
+        seed: random seed to randomly split training, validation and test datasets, defaut is 0.
+        cache_num: number of items to be cached. Default is `sys.maxsize`.
+            will take the minimum of (cache_num, data_length x cache_rate, data_length).
+        cache_rate: percentage of cached data in total, default is 1.0 (cache all).
+            will take the minimum of (cache_num, data_length x cache_rate, data_length).
+        num_workers: the number of worker threads to use.
+            If 0 a single thread will be used. Default is 0.
+
+    """
+
+    resource = "https://www.dropbox.com/s/5wwskxctvcxiuea/MedNIST.tar.gz?dl=1"
+    md5 = "0bc7306e7427e00ad1c5526a6677552d"
+
+    def __init__(
+        self,
+        root: str,
+        section: str,
+        transform: Callable[..., Any],
+        download: bool = False,
+        extract: bool = False,
```

Comment:
I think we always want `extract=True` , perhaps remote this arg?

Reply:
1. "download = True and extract = True" is to download dataset tar and extract.
2. "download = False and extract = True" is for the case that dataset tar exists but not extracted(maybe manually copied from other places).
3. "download = False and extract = False" is for the case that the extracted dataset exists.

What do you think?

Reply:
I think this class is in the application folder, so I assume it's a relatively high-level API. Downloading/unzipping are implementation details. the user shouldn't need to care about whether the data is stored in a tar file and the decompressing process at this level of API

Reply:
I think it's very common for users to copy dataset file instead of downloading it, do you mean to still call `extract` and skip extracting if dataset folder exists?

Reply:
Yes, skip extracting if the folder exists. Because there are datasets in lmdb format which doesn't require an "extract" step. Users of `MedNISTDataset` may not know whether MedNIST is originally stored in tar or lmdb format.

If we provide `extract=True`, then I think we probably need to provide `source_tarfile_name="myfolder/myfile.tar.gz"`, we could do that mid-level API in `monai.data` but not in `monai.application`

Reply:
OK, updated to use `download_and_extract` in the latest commit.